### PR TITLE
[dynamo] add break_graph_if_unsupported to LOAD_ATTR

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #189602
* #187707
* __->__ #187468
* #191059
* #191030
* #187532
* #191963

LOAD_ATTR previously relied on the step() fallback for graph breaks,
which restarts from the most recent empty-stack checkpoint. This is
less precise than other opcodes like STORE_ATTR and CALL that use
break_graph_if_unsupported to break at the exact instruction boundary.

Add the decorator to the non-method LOAD_ATTR path (push=True) so graph
breaks compile a partial graph at the LOAD_ATTR boundary instead of
restarting from the previous empty-stack checkpoint, matching STORE_ATTR
and CALL. The method path (LOAD_METHOD, Python 3.12+ LOAD_ATTR with
arg % 2) still uses the step() fallback because it pushes 2 values
(NULL + method), which the decorator can't handle (it only supports 0
or 1 outputs).

Compiling through the LOAD_ATTR boundary now traces code paths that the
old frame-skip behavior never reached, exposing several latent bugs.
The supporting fixes:

1. Guard tensor.grad_fn when it constant-folds to None. For a leaf
   tensor, grad_fn is None and method_attr_grad_fn returns a constant
   None. requires_grad alone does not distinguish a leaf (grad_fn is
   None) from a non-leaf with identical metadata, and TENSOR_MATCH does
   not guard grad_fn presence, so a first call with a leaf baked None
   in and a later non-leaf call reused it with no recompile. Install a
   NONE_MATCH guard on tensor.grad_fn so a non-leaf input recompiles.
   This is a pre-existing bug, previously masked because grad_fn access
   frame-skipped the whole frame to eager.

2. Graph-break on SavedVariable-backed _saved_* getset access on autograd
   Nodes. Those descriptors' __get__ unpacks the saved tensor, which fires
   non-reentrant-checkpoint recomputation hooks, so running the getter at
   trace time is observable. Worse, the trace-time read also installs a
   source-backed guard over the attribute, and the C++ guard re-runs the
   getter on every evaluation: a checkpointed region measured 12
   recomputations where eager does 2.

   The break is raised with unimplemented(), NOT with
   _UnhandledDescriptorError -- an intermediate version of this change used
   the latter and was wrong. _UnhandledDescriptorError means "caller, fall
   back", not "graph break": VariableTracker.tp_getattro_impl catches it and
   degrades to NotImplementedError, which generic_getattr turns into a
   deferred GetAttrVariable. A deferred attribute access is materialized only
   where the value is reconstructed, so for a discarded result --
   `node._saved_result` as a bare statement, or a dead local -- the access is
   never emitted and the recomputation silently never happens at all. That
   traded the 12-instead-of-2 overcount for a 1-instead-of-2 undercount
   (test_access_saved_tensor_twice_without_recomputation_works). A deferred
   VT is only sound for a pure getter; a getter with observable side effects
   has to graph-break.

   The predicate is narrowed to SavedVariable-backed fields via a
   `_raw_saved_<name>` sibling probe. The autograd codegen emits that sibling
   iff the field is a SavedVariable (should_append_raw_getsetdef in
   tools/autograd/gen_autograd_functions.py), verified exhaustively over the
   1834 generated _saved_* getters with zero mismatches in either direction.
   Without the narrowing, plain C-struct reads such as _saved_dim and
   _saved_self_sym_sizes -- which have no side effect and used to
   constant-fold -- would graph-break too, and a read inside a loop
   escalates to a whole-frame skip.

3. Catch SpeculationLogDivergence in run_tracer() and convert to
   SkipFrame. When a restarted analysis diverges (e.g. due to a
   skip_frame() in an inlined function), the divergence was previously
   unhandled; now it gracefully falls back to eager.

4. Add gc.collect() in TestNestedCheckpoint.check_graph_dies. Dynamo's
   compiled code holds references (compiled fn -> FX graph -> example
   values) that prevent immediate refcount-based GC of autograd nodes.
   The test relied on refcount GC; gc.collect() reclaims them.

Two further latent bugs this exposes are fixed by the two commits this
one now sits on top of, rather than by later commits in the stack, so
that this PR is green on its own: the autograd.Function
non-tensor-output arity bug ("preserve non-tensor autograd.Function
outputs in reconstructed node") and the custom-__del__ object
reconstruction bug ("graph break when constructing an object with a
custom __del__"). With those, no dynamo_wrapped autograd tests need to
be marked as expected failures.

Review order: read the symbolic_convert.py decorator first, then the
four supporting fixes, then the tests.

Test Plan:

Regressions this change originally introduced in the dynamo_wrapped CI
job, all three now green at this commit:

```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_autograd.py \
    TestAutograd.test_access_saved_tensor_twice_without_recomputation_works \
    TestAutograd.test_inplace_on_view_saved_output \
    TestAutograd.test_grad_fn_prehooks_multiple_outputs
# Ran 3 tests ... OK
```

The first of those is a direct metric for fix 2: it counts checkpoint
recomputations, and went 1 -> 2 (expected 2) with the unimplemented()
break in place.

```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_autograd.py TestAutograd
python test/dynamo/test_decorators.py
python test/dynamo/test_tp_getattro.py
python test/dynamo/test_sets.py
python test/dynamo/test_misc.py
python test/dynamo/test_functions.py
python test/dynamo/test_ctx_manager.py
python test/dynamo/test_repros.py
python test/dynamo/test_activation_checkpointing.py
```

All remaining errors were verified to fail identically on the unpatched
merge-base (the local build lacks cuDNN), so this introduces no
regressions.

Co-authored-by: Claude <noreply@anthropic.com>

### CODE REVIEW

**Review by guilhermeleobas — APPROVED** [review]

**Comment by williamwen42** [comment]
NOTE: benchmark graph break numbers are up because we're no longer skipping code due to graph break in LOAD_ATTR - we're tracing through more code, which results in more graph breaks.
